### PR TITLE
Isuues with Gson Jackson Bridge

### PR DESCRIPTION
**Pull request by bmazzarol**
- Fixed issue where integers are always printed as doubles
- Fixed issue using the jackson XML streaming parser where boolean tokens are always strings

### CODE REVIEW

**Comment by bmazzarol** [comment]
Hi Guys,

Thanks for the great library.

I found a few issues using other Jackson data formats (xml, smile, yaml) using the Gson/Jackson bridge.

This resolved the two issues listed above.

Happy to make any changes/add any tests, provide more information, etc...

Cheers
Ben

File: gson/src/org/immutables/gson/stream/JsonParserReader.java [hunk 1]
```diff
@@ -143,7 +144,12 @@ public String nextString() throws IOException {
   @Override
   public boolean nextBoolean() throws IOException {
     requirePeek();
-    boolean value = parser.getBooleanValue();
+    boolean value;
+    try {
+      value = parser.getBooleanValue();
```
Comment (asereda-gs) [inline gson/src/org/immutables/gson/stream/JsonParserReader.java]: Can you use `parser.getCurrentToken()` to detect JSON type ?

**Comment by asereda-gs** [comment]
Hi Ben,

Thanks for this PR. 

Can you please explain why do you need to write separately  ints and doubles ? What is the usecase ?  

```java
// pseudo-code (simplified logic)
if (isInteger(a)) {
  writer.writeInt(a);
} else {
  writer.writeDouble(a);
}
```

Also is it possible to add tests for this change ?

File: gson/src/org/immutables/gson/stream/JsonParserReader.java (hunk 1 above)
Comment (asereda-gs) [reply gson/src/org/immutables/gson/stream/JsonParserReader.java]: Also what is the default behavior of GSON `JsonReader.nextBoolean()`:

```json
{ 
  "foo": true, 
  "bar": "true"
} 
```

will it return `true` in both cases ?

File: gson/src/org/immutables/gson/stream/JsonGeneratorWriter.java [hunk 2]
```diff
@@ -145,7 +147,12 @@ public JsonWriter value(Number value) throws IOException {
         throw new IllegalArgumentException("JSON forbids NaN and infinities: " + value);
       }
     }
-    generator.writeNumber(d);
+    if (floor(d) == d) {
```
Comment (asereda-gs) [inline gson/src/org/immutables/gson/stream/JsonGeneratorWriter.java]: I think it is better to perform `instanceof`:

```java
if (value instanceof Integer) {
  generator.writeNumber(value.intValue());
} else if (value instanceof Float) {
  generator.writeNumber(value.floatNumber());
} else if (value instanceof Short) {
  generator.writeNumber(value.shortValue());
} else if{
//.... Byte / BigInteger / BigDecimal etc.. 
} else {
  // use double by default
  writer.writeNumber(value.doubleValue());
}
```

File: gson/src/org/immutables/gson/stream/JsonParserReader.java (hunk 1 above)
Comment (bmazzarol) [reply gson/src/org/immutables/gson/stream/JsonParserReader.java]: This is not an issue for Gson or the Jackson JsonParser. They both correctly distinguish the token type as boolean. The issue is with the Jackson XML JsonParser. All boolean tokens come come across as a string tokens. So calling `JsonReader.nextBoolean()` Fails here https://github.com/FasterXML/jackson-core/blob/2.11/src/main/java/com/fasterxml/jackson/core/JsonParser.java#L1426-L1431

File: gson/src/org/immutables/gson/stream/JsonGeneratorWriter.java (hunk 2 above)
Comment (bmazzarol) [reply gson/src/org/immutables/gson/stream/JsonGeneratorWriter.java]: The reason is because strict parsers that are expecting an integer will fail if a double is found instead. And because it causes a deviation of the output from what Gson produces verses what this produces.

The reason I did not do the instance of checks is because I did not want to have to account for all the possible sub-types of Number. And I did not think the performance would be adversely effected by one call to floor. By my count there is at least 18 classes that directly extend Number in the standard lib, not to mention all the siblings from there.

I will add some unit tests and do a performance comparison between the instance of checks and the comparison with floor.

Comment (asereda-gs) [reply gson/src/org/immutables/gson/stream/JsonGeneratorWriter.java]: I still think it is better to do `instanceof`. For example jackson has special methods for  `BigInteger` / `BigDecimal` for not loosing number precision it is better to leverage them.

Don't forget you're not only doing `Math.floor()` but also `number.doubleValue()` instead of single `instanceof`.

File: gson/src/org/immutables/gson/stream/JsonParserReader.java (hunk 1 above)
Comment (asereda-gs) [reply gson/src/org/immutables/gson/stream/JsonParserReader.java]: `JsonReader.nextBoolean()` should only be called when `currentToken` is `VALUE_TRUE` or  `VALUE_FALSE`. 

Perhaps something is wrong elsewhere (like method `getCurrentToken()`) ?

**Comment by bmazzarol** [comment]
@asereda-gs used instanceof checks for all supported Number subtypes, with a catch all that is still integer aware, (uses floor in this case).

I have added unit tests for all the schema-free formats that Jackson has (minus Ion because it is only supported at 2.10). To get serialization working for xml and properties formats I had to add a new reader that expects the producer to only produce string tokens, try using those formats with the JsonParserReader to see the issue.

File: gson/src/org/immutables/gson/stream/JsonGeneratorWriter.java [hunk 3]
```diff
@@ -145,7 +150,25 @@ public JsonWriter value(Number value) throws IOException {
         throw new IllegalArgumentException("JSON forbids NaN and infinities: " + value);
       }
     }
-    generator.writeNumber(d);
+    if (value instanceof Integer) {
+      generator.writeNumber((Integer) value);
```
Comment (asereda-gs) [inline gson/src/org/immutables/gson/stream/JsonGeneratorWriter.java]: you can use `value.intValue()` here.
The implementation of this method will return `this` (for Integer)

File: gson/src/org/immutables/gson/stream/JsonGeneratorWriter.java [hunk 4]
```diff
@@ -145,7 +150,25 @@ public JsonWriter value(Number value) throws IOException {
...
+    } else if (value instanceof Short) {
+      generator.writeNumber((Short) value);
```
Comment (asereda-gs) [inline gson/src/org/immutables/gson/stream/JsonGeneratorWriter.java]: same `value.shortValue()` here

File: gson/src/org/immutables/gson/stream/JsonGeneratorWriter.java [hunk 5]
```diff
@@ -145,7 +150,25 @@ public JsonWriter value(Number value) throws IOException {
...
+      generator.writeNumber((Long) value);
+    } else if (value instanceof Float) {
+      generator.writeNumber((Float) value);
+    } else if (value instanceof Double) {
+      generator.writeNumber((Double) value);
+    } else if (value instanceof BigInteger) {
+      generator.writeNumber((BigInteger) value);
+    } else if (value instanceof BigDecimal) {
+      generator.writeNumber((BigDecimal) value);
+    } else if (floor(d) == d) {
```
Comment (asereda-gs) [inline gson/src/org/immutables/gson/stream/JsonGeneratorWriter.java]: Please move the [statement](https://github.com/immutables/immutables/blob/9d6392258ce5b47b09d7d6b35a5f5b168de3471b/gson/src/org/immutables/gson/stream/JsonGeneratorWriter.java#L142) `double d = value.doubleValue()` (not visible in this review) to where conversion to double is necessary. For instance you don't need conversion to double when you know that current type is `Integer`.

Comment (asereda-gs) [inline gson/src/org/immutables/gson/stream/JsonGeneratorWriter.java]: Let me think more about cases when a double / float `1.0D` needs to be serialized as int `1`. 

One might potentially lose information like positive and negative zero `-0.0F` and `+0.0F`

File: gson/pom.xml [hunk 6]
```diff
@@ -163,6 +200,14 @@
           </execution>
         </executions>
       </plugin>
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-compiler-plugin</artifactId>
+        <configuration>
+          <source>8</source>
```
Comment (asereda-gs) [inline gson/pom.xml]: Can we keep java 7 ? Is there any reason to make it java 8 ? 

I'm not sure if this module is being used by Android developers.

File: gson/test/org/immutables/gson/bridge/GsonJacksonBridgeSerializationTest.java [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+              try {
+                return factory.createParser(os.toByteArray());
+              } catch (IOException e) {
+                throw new RuntimeException(e);
+              }
+            }
+            , JsonParserReader::new);
+  }
+
+  private <VALUE
```
Comment (asereda-gs) [inline gson/test/org/immutables/gson/bridge/GsonJacksonBridgeSerializationTest.java]: This method is hard to follow (5 arguments and 3 generic types). 
Do you need separate `WRITER` / `READER` generic types ? Can one keep them as subclasses to `JsonGenerator` / `JsonParser`

File: gson/pom.xml (hunk 6 above)
Comment (bmazzarol) [reply gson/pom.xml]: Sure

File: gson/src/org/immutables/gson/stream/JsonGeneratorWriter.java (hunk 4 above)
Comment (bmazzarol) [reply gson/src/org/immutables/gson/stream/JsonGeneratorWriter.java]: Sure

File: gson/src/org/immutables/gson/stream/JsonGeneratorWriter.java (hunk 5 above)
Comment (bmazzarol) [reply gson/src/org/immutables/gson/stream/JsonGeneratorWriter.java]: What about the strict validation you were performing?

Comment (bmazzarol) [reply gson/src/org/immutables/gson/stream/JsonGeneratorWriter.java]: More than happy for the else branch to be double. Let me know what you want here

File: gson/test/org/immutables/gson/bridge/GsonJacksonBridgeSerializationTest.java (hunk 7 above)
Comment (bmazzarol) [reply gson/test/org/immutables/gson/bridge/GsonJacksonBridgeSerializationTest.java]: Sure can re-write it.

File: gson/src/org/immutables/gson/stream/JsonGeneratorWriter.java (hunk 5 above)
Comment (asereda-gs) [reply gson/src/org/immutables/gson/stream/JsonGeneratorWriter.java]: Yes.  Right now I think double should be last resort (if no other types are matching) :
```java
// many instanceof checks 
} else {
  // last resort serialize as double
 generator.writeNumber(number.doubleValue());
}

```

Comment (bmazzarol) [reply gson/src/org/immutables/gson/stream/JsonGeneratorWriter.java]: Left the `double d = value.doubleValue()` in there to enable isNan strict checks

File: gson/src/org/immutables/gson/stream/JsonGeneratorWriter.java [hunk 8]
```diff
@@ -145,7 +148,23 @@ public JsonWriter value(Number value) throws IOException {
...
+    } else if (value instanceof Float) {
+      generator.writeNumber(value.floatValue());
+    } else if (value instanceof Double) {
+      generator.writeNumber(value.doubleValue());
+    } else if (value instanceof BigInteger) {
+      generator.writeNumber((BigInteger) value);
+    } else if (value instanceof BigDecimal) {
+      generator.writeNumber((BigDecimal) value);
+    } else {
+      generator.writeNumber(d);
```
Comment (asereda-gs) [inline gson/src/org/immutables/gson/stream/JsonGeneratorWriter.java]: why not `generator.writeNumber(value.doubleValue())` ?

File: gson/src/org/immutables/gson/stream/JsonParserReader.java [hunk 9]
```diff
@@ -143,7 +144,12 @@ public String nextString() throws IOException {
...
+    } catch (JsonParseException e) {
+      value = Boolean.parseBoolean(parser.getValueAsString());
```
Comment (asereda-gs) [inline gson/src/org/immutables/gson/stream/JsonParserReader.java]: Is your proposed test `GsonJacksonBridgeSerializationTest`  failing without try / catch ?

**Comment by asereda-gs** [comment]
Overall LGTM with some clarifications:
1. Parsing boolean from String (after `JsonParseException`). Want to make sure there are no better alternatives. 
2. Unused `d` variable

File: gson/src/org/immutables/gson/stream/JsonParserReader.java (hunk 9 above)
Comment (bmazzarol) [reply gson/src/org/immutables/gson/stream/JsonParserReader.java]: This code can be reverted. It isn't, but is not required as I have subclasses the reader. So to clarify formats that do not support tokens other than string, such as xml and property, will have to use the XmlParserReader.

File: gson/src/org/immutables/gson/stream/JsonGeneratorWriter.java (hunk 8 above)
Comment (bmazzarol) [reply gson/src/org/immutables/gson/stream/JsonGeneratorWriter.java]: Fine but in it will mean in cases of a double value with strict checks, doubleValue will Be called twice

Comment (asereda-gs) [reply gson/src/org/immutables/gson/stream/JsonGeneratorWriter.java]: you're talking about `isLinient()`

```java
} else {
   // remove previous "value instanceof Double"
    double d = value.doubleValue();
    if (!isLenient()) {
      if (Double.isNaN(d) || Double.isInfinite(d)) {
        throw new IllegalArgumentException("JSON forbids NaN and infinities: " + value);
      }
    }
   generator.writeNumber(d);
}
```

File: gson/src/org/immutables/gson/stream/JsonParserReader.java (hunk 9 above)
Comment (asereda-gs) [reply gson/src/org/immutables/gson/stream/JsonParserReader.java]: Let's keep this logic in `XmlParserReader` then. I would like to avoid making too many assumptions in `JsonParserReader` if that's OK with you.

File: gson/src/org/immutables/gson/stream/JsonGeneratorWriter.java [hunk 10]
```diff
@@ -139,13 +142,28 @@ public JsonWriter value(Number value) throws IOException {
     if (value == null) {
       return nullValue();
     }
-    double d = value.doubleValue();
     if (!isLenient()) {
```
Comment (asereda-gs) [inline gson/src/org/immutables/gson/stream/JsonGeneratorWriter.java]: @bmazzarol sorry for being pedantic but what I meant is the following:

```java
   if (value instanceof Integer) {
      generator.writeNumber(value.intValue());
    } else if (value instanceof Short || value instanceof Byte) { // write byte as short
      generator.writeNumber(value.shortValue());
    } else if (value instanceof Long) {
      generator.writeNumber(value.longValue());
    } else if (value instanceof Float) {
      generator.writeNumber(value.floatValue());
   // } else if (value instanceof Double) {
   // remove instanceof Double. it is handled in else clause
   //  generator.writeNumber(value.doubleValue());
    } else if (value instanceof BigInteger) {
      generator.writeNumber((BigInteger) value);
    } else if (value instanceof BigDecimal) {
      generator.writeNumber((BigDecimal) value);
    } else {
       // by default, write as double
      // converting to double here (and only here)
     double d = value.doubleValue();
    if (!isLenient() && (Double.isNaN(d) || Double.isInfinite(d))) {
        throw new IllegalArgumentException("JSON forbids NaN and infinities: " + value + " ( for " + value.getClass() + ")");	   
     }	
    generator.writeNumber(d);	    
    }
```

This way you convert to double only once.

File: gson/src/org/immutables/gson/stream/JsonGeneratorWriter.java [hunk 11]
```diff
@@ -142,9 +142,8 @@ public JsonWriter value(Number value) throws IOException {
...
-      if (Double.isNaN(d) || Double.isInfinite(d)) {
+      if (Double.isNaN(value.doubleValue()) || Double.isInfinite(value.doubleValue())) {
```
Comment (elucash) [inline gson/src/org/immutables/gson/stream/JsonGeneratorWriter.java]: this could be a silly comment on my side, I've noticed that we call `value.doubleValue()` many times, cannot justify the impact because of jit/inlining, just mentioning

Comment (asereda-gs) [reply gson/src/org/immutables/gson/stream/JsonGeneratorWriter.java]: We'll call it only once in `else`.
See my [previous comment](https://github.com/immutables/immutables/pull/1120#discussion_r345316915)

File: gson/src/org/immutables/gson/stream/JsonGeneratorWriter.java (hunk 10 above)
Comment (bmazzarol) [reply gson/src/org/immutables/gson/stream/JsonGeneratorWriter.java]: I know what you are asking for, but it is semantically different from the code you had before this pr, and is not correct. If I pass a Double value that is NaN or infinite and strict parsing is enabled no error will be raised.

File: gson/src/org/immutables/gson/stream/JsonGeneratorWriter.java (hunk 11 above)
Comment (bmazzarol) [reply gson/src/org/immutables/gson/stream/JsonGeneratorWriter.java]: When strict parsing is enabled, You will have to pay the cost of either a cast, or conversion to double for all Number types that can be NaN or infinite. I'm open to how you want to do that, but can't  speculate on what the JVM does performance wise.

File: gson/src/org/immutables/gson/stream/JsonGeneratorWriter.java (hunk 10 above)
Comment (asereda-gs) [reply gson/src/org/immutables/gson/stream/JsonGeneratorWriter.java]: > If I pass a Double value that is NaN or infinite and strict parsing is enabled no error will be raised.

Would it not get into `else` statement and throw exception ? 

Or are you talking about the case when `BigDecimal.doubleValue()` returns `-Infinity` / `+Infinity` ?

File: gson/src/org/immutables/gson/stream/JsonGeneratorWriter.java (hunk 11 above)
Comment (asereda-gs) [reply gson/src/org/immutables/gson/stream/JsonGeneratorWriter.java]: Why not check only for Floats and Doubles (BigDecimal is a special case) when strict checking is enabled ? 

```java
    System.out.printf("POSITIVE_INFINITY: %s%n", Double.isInfinite(Float.POSITIVE_INFINITY));
    System.out.printf("NEGATIVE_INFINITY: %s%n", Double.isInfinite(Float.NEGATIVE_INFINITY));
    System.out.printf("NaN: %s%n", Double.isNaN(Float.NaN));
```

> POSITIVE_INFINITY: true
> NEGATIVE_INFINITY: true
> NaN: true

Comment (asereda-gs) [reply gson/src/org/immutables/gson/stream/JsonGeneratorWriter.java]: I would say have private method:

```java

public JsonWriter value(Number value) {

   if (value instanceof Float) {
     float f = value.floatValue();
     checkStrictNumber(f);
     generator.writeNumber(f);	
   } else {
      double d = value.doubleValue();
      checkStrictNumber(d);
      generator.writeNumber(d);	
   }
}

void checkStrictNumber(double number) {
  if (isLinient() && (Double.isInfinite(number) || Double.isNaN(number))) {
    throw new IllegalArgumentException("JSON forbids NaN and infinities: " + value);
  }
} 
```

Do you agree with this approach ?

Comment (bmazzarol) [reply gson/src/org/immutables/gson/stream/JsonGeneratorWriter.java]: Yep will implement.

**Comment by asereda-gs** [comment]
Thank you @bmazzarol for your contribution!
